File: code/0817 Linked List Components.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
int numComponents(struct ListNode* head, int* G, int GSize) {
    int sum=0;
    bool isConnect=false;
    int *map=(int*)calloc(10001,sizeof(int));
    for(int i=0;i<GSize;i++){
        map[G[i]]++;
    }
    while(head){
        
        if(head->val<10001&&map[head->val]!=0){
            if(!isConnect){
                sum++;
                isConnect=true;
            }
        }else{
            isConnect=false;
        }
        head=head->next;
    }
    return sum;
}

File: code/0817 Linked List Components.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def numComponents(self, head, G):
        """
        :type head: ListNode
        :type G: List[int]
        :rtype: int
        """
        count = p = 0
        G = set(G)
        while head:
            if head.val in G:
                if p == 0:
                    count += 1
                    p = 1
            else:
                p = 0
            head = head.next
        return count
        
